File: apps/backend/src/ev_twin_api/schemas/optimization.py
from pydantic import BaseModel, ConfigDict, Field, model_validator

from ev_twin_api.schemas.scenario import Scenario


class LayoutCandidate(BaseModel):
    layout_id: str = Field(min_length=1, max_length=80)
    layout_version: int = Field(ge=1)


class OptimizationRequest(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)

    name_prefix: str = Field(min_length=1, max_length=50)
    layouts: list[LayoutCandidate] = Field(min_length=1, max_length=8)
    route_ids: list[str] = Field(min_length=1, max_length=8)
    robot_counts: list[int] = Field(min_length=1, max_length=8)
    robot_speeds_mps: list[float] = Field(min_length=1, max_length=8)
    charger_counts: list[int] = Field(min_length=1, max_length=8)
    demand_intervals: list[float] = Field(min_length=1, max_length=8)
    num_tasks: int = Field(default=100, ge=1, le=10_000)
    loading_time: float = Field(default=5.0, gt=0.0, le=86_400.0)
    simulation_time: float = Field(default=3_600.0, gt=0.0, le=86_400.0)

    @model_validator(mode="after")
    def bounded_search(self) -> "OptimizationRequest":
        dimensions = (
            self.layouts,
            self.route_ids,
            self.robot_counts,
            self.robot_speeds_mps,
            self.charger_counts,
            self.demand_intervals,
        )
        candidate_count = 1
        for dimension in dimensions:
            candidate_count *= len(dimension)
        if candidate_count > 64:
            raise ValueError("optimization search is limited to 64 candidates")
        if any(not 1 <= value <= 10 for value in self.robot_counts):
            raise ValueError("robot_counts values must be in [1, 10]")
        if any(not 0.0 < value <= 10.0 for value in self.robot_speeds_mps):
            raise ValueError("robot_speeds_mps values must be in (0, 10]")
        if any(not 1 <= value <= 20 for value in self.charger_counts):
            raise ValueError("charger_counts values must be in [1, 20]")
        if any(not 1.0 <= value <= 60.0 for value in self.demand_intervals):
            raise ValueError("demand_intervals values must be in [1, 60]")
        return self


class RankedScenario(BaseModel):
    rank: int = Field(ge=1)
    scenario: Scenario


class OptimizationResult(BaseModel):
    evaluated_candidates: int = Field(ge=1, le=64)
    recommendation: Scenario
    ranking: list[RankedScenario]

File: apps/frontend/src/schemas/optimization.ts
import { z } from "zod";
import { scenarioSchema } from "./scenario";

export const layoutCandidateSchema = z.object({
  layout_id: z.string().min(1).max(80),
  layout_version: z.number().int().min(1),
});

interface OptimizationDimensions {
  layouts: unknown[];
  route_ids: unknown[];
  robot_counts: unknown[];
  robot_speeds_mps: unknown[];
  charger_counts: unknown[];
  demand_intervals: unknown[];
}

export function optimizationCandidateCount(request: OptimizationDimensions) {
  return [
    request.layouts,
    request.route_ids,
    request.robot_counts,
    request.robot_speeds_mps,
    request.charger_counts,
    request.demand_intervals,
  ].reduce((count, dimension) => count * dimension.length, 1);
}

export const optimizationRequestSchema = z.object({
  name_prefix: z.string().trim().min(1).max(50),
  layouts: z.array(layoutCandidateSchema).min(1).max(8),
  route_ids: z.array(z.string().min(1).max(80)).min(1).max(8),
  robot_counts: z.array(z.number().int().min(1).max(10)).min(1).max(8),
  robot_speeds_mps: z.array(z.number().positive().max(10)).min(1).max(8),
  charger_counts: z.array(z.number().int().min(1).max(20)).min(1).max(8),
  demand_intervals: z.array(z.number().min(1).max(60)).min(1).max(8),
  num_tasks: z.number().int().min(1).max(10_000).default(100),
  loading_time: z.number().positive().max(86_400).default(5),
  simulation_time: z.number().positive().max(86_400).default(3_600),
}).superRefine((request, context) => {
  const candidates = optimizationCandidateCount(request);
  if (candidates > 64) {
    context.addIssue({ code: "custom", message: "Optimization is limited to 64 candidates" });
  }
});

export const rankedScenarioSchema = z.object({
  rank: z.number().int().min(1),
  scenario: scenarioSchema,
});

export const optimizationResultSchema = z.object({
  evaluated_candidates: z.number().int().min(1).max(64),
  recommendation: scenarioSchema,
  ranking: z.array(rankedScenarioSchema),
});

export type OptimizationRequest = z.input<typeof optimizationRequestSchema>;
export type OptimizationResult = z.infer<typeof optimizationResultSchema>;
